scipy/cluster/hierarchy.py

```diff
@@ -1395,6 +1400,208 @@ def cut_tree(Z, n_clusters=None, height=None):
+    as the beginning. Further, this data structure is used in order to perform
+    the successive computations.
+
+
+    Examples
+    --------
+    >>> from scipy.cluster import hierarchy
+    >>> from scipy import stats
+
+    Initialize the random seed.
+
+    >>> np.random.seed(14)
+
+    Create a input matrix containing 100 data samples with 4 dimensions.
+    Note: using `gamma` in order to generate an unbalanced distribution.
+    If a regular ``cut_tree()`` would be performed, one big and many small
+    clusters would be obtained.
+
+    >>> X = stats.gamma.rvs(0.1, size=400).reshape((100, 4))
+
+    Compute the linkage matrix using the scipy ward() or linkage() method:
+
+    >>> Z = hierarchy.ward(X)
+
+    Perform a balanced cut tree of the linkage matrix:
+
+    >>> cluster_id, cluster_level = hierarchy.cut_tree_balanced(
+    ...                                              Z, max_cluster_size=10)
+    >>> cluster_id[:10]
+    array([19,  4, 10, 12, 20, 12, 14,  9, 15,  2])
+    >>> cluster_level[:10]
+    array([array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 1, 0])], dtype=object)
+
+    Note that clusters with more similar values for ``cluster_level`` denote
+    clusters with less distance in between, thus representing vectors which are
+    closer in the multidimensional space. This information contained within
+    ``cluster_level`` is not usually present in ``cluster_id``.
+
+    """
+    # Assert that the input max_cluster_size is >= 1
+    if not max_cluster_size >= 1:
+        raise ValueError("max_cluster_size should be >= 1, is: {}"
+                         .format(max_cluster_size))
+
+    # Perform a full cut tree of the linkage matrix
+    full_cut = cut_tree(Z)
+
+    # Initialize the variable containing the current cluster id (it will be
+    # higher for each newly found valid cluster)
+    last_cluster_id = 0
+
+    # Initialize the resulting cluster id vector (containing for each row in
+    # input_data_x_sample its corresponding cluster id)
+    ndim = full_cut.shape[1]
+    cluster_id = np.zeros(ndim, dtype=int)
+
+    # Initialize the resulting cluster level vector (containing for each data
+    # sample its corresponding cluster tree level)
+    cluster_level = np.empty((ndim,), dtype=object)
+    for i in range(ndim):
+        cluster_level[i] = np.array([0], int)
+
+    # Scan the full cut matrix from the last column (root tree level) to the
+    # first column (leaves tree level)
+    for icol in range(ndim-1, -1, -1):
```

Comment:
probably minor--I was just scanning through the codecov patch diff, and noticed a partial hit on this line--not sure if there are sensible tests that would exercise it more fully (maybe the iteration can in rare cases go farther than is probed by the tests?).

Reply:
Many thanks for this sensible suggestion. I added some edge test cases in order to cover these possible issues.
Note that `ndim` represents the number of dimensions of the full cut, which is computed using the input linkage matrix.  Note also that the [function to compute the linkage matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html) never returns a matrix with 0 or 1 values, so the input `Z` at `cut_tree_balanced` is not expected to be empty or of size 1. 
Anyhow, in case we receive an invalid linkage matrix as input (i.e. empty or of size 1), the `cut_tree()` function (used to compute the full cut) will throw a `ValueError` (which is asserted in the new test cases).
As a corollary, the `full_cut` matrix will be always at least of 2x2 size, and therefore `ndim` is ensured to be `>=0`.
Please let me know if further test cases should be added.
Many thanks and best regards.